resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java

```diff
@@ -0,0 +1,319 @@
+                if (entries != null) {
+                    entries.clearAndCancel();
+                }
+            }
+        }, ttl, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public String toString() {
+        return new StringBuilder()
+                .append("DefaultAuthoritativeDnsServerCache(minTtl=")
+                .append(minTtl).append(", maxTtl=")
+                .append(maxTtl).append(", cached nameservers=")
+                .append(resolveCache.size()).append(")")
+                .toString();
+    }
+
+    private static final class Entry {
+
+        volatile ScheduledFuture<?> expirationFuture;
+
+        private final InetSocketAddress address;
+
+        Entry(InetSocketAddress address) {
+            this.address = checkNotNull(address, "address");
+        }
+
+        InetSocketAddress nameServer() {
+            return address;
+        }
+
+        void scheduleExpiration(EventLoop loop, Runnable task, long delay, TimeUnit unit) {
+            assert expirationFuture == null : "expiration task scheduled already";
+            expirationFuture = loop.schedule(task, delay, unit);
+        }
+
+        void cancelExpiration() {
+            ScheduledFuture<?> expirationFuture = this.expirationFuture;
+            if (expirationFuture != null) {
+                expirationFuture.cancel(false);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return address.toString();
+        }
+    }
+
+    // This List implementation always has RandomAccess semantics as we either wrap an ArrayList, empty List or
+    // one which has only one element.
+    private static final class EntryList extends AbstractList<InetSocketAddress> implements RandomAccess {
+
+        static final EntryList EMPTY = new EntryList(Collections.<Entry>emptyList());
+
+        private final List<Entry> entries;
+
+        EntryList(Entry entry) {
+            this(Collections.singletonList(entry));
+        }
+
+        EntryList(List<Entry> entries) {
+            this.entries = entries;
+        }
+
+        @Override
+        public InetSocketAddress get(int i) {
+            return entries.get(i).address;
+        }
+
+        @Override
+        public int size() {
+            return entries.size();
+        }
+
+        Entry getEntry(int i) {
+            return entries.get(i);
+        }
+
+        boolean cancelExpiration() {
+            final int numEntries = size();
+            if (numEntries == 0) {
+                return false;
+            }
+            for (int i = 0; i < numEntries; i++) {
+                entries.get(i).cancelExpiration();
+            }
+            return true;
+        }
+    }
+
+    // Directly extend AtomicReference for intrinsics and also to keep memory overhead low.
+    private static final class Entries extends AtomicReference<EntryList> {
+
+        Entries(Entry entry) {
+            super(new EntryList(entry));
+        }
+
+        Entry add(Entry e) {
```

Comment:
Why not just use CopyOnWriteArrayList ?

Reply:
@carl-mastrangelo let me think about it... I just followed what we have in `DefaultDnsCache`.

Reply:
Actual I think the current approach is better... If we find out that `CopyOnWriteArrayList` is better we can still do a followup. For example you could do a PR if you think it worth it... WDYT @carl-mastrangelo

Reply:
It doesn't seem like a big issue.   You can leave it as is.